<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


View this project on GitHub (https://github.com/JS-24693/Analyzing-Historical-Stock-Revenue-Data-and-Building-a-Dashboard.git)

<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working Locally using anaconda, please uncomment the following code and execute it.
Use the version as per your python version.


In [1]:
!pip install yfinance
!pip install bs4
!pip install nbformat

In [2]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [3]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. **You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.**


In [5]:
def make_graph(stock_data, revenue_data, stock): #Function to create graph
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing=0.3)
    
    # Filter data to only include dates up to 2021-06-14
    stock_data_specific = stock_data[stock_data.Date <= '2021-06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']

    # Add traces for stock data and revenue data
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    # Update x and y axis titles
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    # Update layout
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    # Show the figure
    fig.show() #Function to show graph

Use the make_graph function that we’ve already defined. You’ll need to invoke it in questions 5 and 6 to display the graphs and create the dashboard. 
> **Note: You don’t need to redefine the function for plotting graphs anywhere else in this notebook; just use the existing function.**


## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [41]:
tesla = yf.Ticker("TSLA") #Get Tesla stock data

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


In [42]:
tesla_data = tesla.history(period="max") #Get Tesla share price data
tesla_data.head() #Get Tesla revenue data

,Open,High,Low,Close,Volume,Dividends,Stock Splits
Date,,,,,,,
2010-06-29 00:00:00-04:00,1.266667,1.666667,1.169333,1.592667,281494500,0.0,0.0
2010-06-30 00:00:00-04:00,1.719333,2.028000,1.553333,1.588667,257806500,0.0,0.0
2010-07-01 00:00:00-04:00,1.666667,1.728000,1.351333,1.464000,123282000,0.0,0.0
2010-07-02 00:00:00-04:00,1.533333,1.540000,1.247333,1.280000,77097000,0.0,0.0
2010-07-06 00:00:00-04:00,1.333333,1.333333,1.055333,1.074000,103003500,0.0,0.0


**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [43]:
tesla_data.reset_index(inplace=True) #reset the index
tesla_data.head() #Get Tesla revenue data

,Date,Open,High,Low,Close,Volume,Dividends,Stock Splits
0,2010-06-29 00:00:00-04:00,1.266667,1.666667,1.169333,1.592667,281494500,0.0,0.0
1,2010-06-30 00:00:00-04:00,1.719333,2.028000,1.553333,1.588667,257806500,0.0,0.0
2,2010-07-01 00:00:00-04:00,1.666667,1.728000,1.351333,1.464000,123282000,0.0,0.0
3,2010-07-02 00:00:00-04:00,1.533333,1.540000,1.247333,1.280000,77097000,0.0,0.0
4,2010-07-06 00:00:00-04:00,1.333333,1.333333,1.055333,1.074000,103003500,0.0,0.0


## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [44]:
# Fetch the data from the URL
url_tesla = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
html_data = requests.get(url_tesla).text

Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


In [45]:
# Parse the HTML data
soup = BeautifulSoup(html_data, 'html5lib')

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Step-by-step instructions</summary>

```

Here are the step-by-step instructions:

1. Create an Empty DataFrame
2. Find the Relevant Table
3. Check for the Tesla Quarterly Revenue Table
4. Iterate Through Rows in the Table Body
5. Extract Data from Columns
6. Append Data to the DataFrame

```
</details>


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1

We are focusing on quarterly revenue in the lab.
```

</details>


In [46]:
# Initialize an empty DataFrame to store the Tesla Revenue table
tesla_revenue = pd.DataFrame(columns=['Date', 'Revenue'])

tables= soup.find_all('table') #finds all the tables in the HTML

for table in tables: #finds all the tables in the HTML
    if 'Tesla Quarterly Revenue' in str(table): #checks if the table contains the text 'Tesla Quarterly Revenue'
        print("Found the Tesla Quarterly Revenue table.")
        # Extract rows from the table
        for row in table.find("tbody").find_all("tr"): #finds the first <tbody> and then all the rows in it
            col = row.find_all("td") #finds all table cells within each row
            date = col[0].text.strip() # gets the date from the first column
            revenue = col[1].text.strip() # gets the revenue from the second column 
    
            # Append the data to the DataFrame
            tesla_revenue = pd.concat([tesla_revenue, pd.DataFrame({"Date":[date],"Revenue":[revenue]})], ignore_index=True)
    

Found the Tesla Quarterly Revenue table.


Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [47]:
# Clean the revenue data
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"", regex=True) #remove the $ sign and commas from the revenue data
# Convert the 'Date' column to datetime format without infer_datetime_format argument
tesla_revenue['Date'] = pd.to_datetime(tesla_revenue['Date'])
#print the cleaned DataFrame
print(tesla_revenue)

         Date Revenue
0  2022-09-30   21454
1  2022-06-30   16934
2  2022-03-31   18756
3  2021-12-31   17719
4  2021-09-30   13757
5  2021-06-30   11958
6  2021-03-31   10389
7  2020-12-31   10744
8  2020-09-30    8771
9  2020-06-30    6036
10 2020-03-31    5985
11 2019-12-31    7384
12 2019-09-30    6303
13 2019-06-30    6350
14 2019-03-31    4541
15 2018-12-31    7226
16 2018-09-30    6824
17 2018-06-30    4002
18 2018-03-31    3409
19 2017-12-31    3288
20 2017-09-30    2985
21 2017-06-30    2790
22 2017-03-31    2696
23 2016-12-31    2285
24 2016-09-30    2298
25 2016-06-30    1270
26 2016-03-31    1147
27 2015-12-31    1214
28 2015-09-30     937
29 2015-06-30     955
30 2015-03-31     940
31 2014-12-31     957
32 2014-09-30     852
33 2014-06-30     769
34 2014-03-31     621
35 2013-12-31     615
36 2013-09-30     431
37 2013-06-30     405
38 2013-03-31     562
39 2012-12-31     306
40 2012-09-30      50
41 2012-06-30      27
42 2012-03-31      30
43 2011-12-31      39
44 2011-09

Execute the following lines to remove an null or empty strings in the Revenue column.


In [48]:
# Remove rows with null values
tesla_revenue.dropna(inplace=True)
# Remove rows with empty strings in the 'Revenue' column
tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]
# Print the cleaned DataFrame
print(tesla_revenue)

         Date Revenue
0  2022-09-30   21454
1  2022-06-30   16934
2  2022-03-31   18756
3  2021-12-31   17719
4  2021-09-30   13757
5  2021-06-30   11958
6  2021-03-31   10389
7  2020-12-31   10744
8  2020-09-30    8771
9  2020-06-30    6036
10 2020-03-31    5985
11 2019-12-31    7384
12 2019-09-30    6303
13 2019-06-30    6350
14 2019-03-31    4541
15 2018-12-31    7226
16 2018-09-30    6824
17 2018-06-30    4002
18 2018-03-31    3409
19 2017-12-31    3288
20 2017-09-30    2985
21 2017-06-30    2790
22 2017-03-31    2696
23 2016-12-31    2285
24 2016-09-30    2298
25 2016-06-30    1270
26 2016-03-31    1147
27 2015-12-31    1214
28 2015-09-30     937
29 2015-06-30     955
30 2015-03-31     940
31 2014-12-31     957
32 2014-09-30     852
33 2014-06-30     769
34 2014-03-31     621
35 2013-12-31     615
36 2013-09-30     431
37 2013-06-30     405
38 2013-03-31     562
39 2012-12-31     306
40 2012-09-30      50
41 2012-06-30      27
42 2012-03-31      30
43 2011-12-31      39
44 2011-09

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [49]:
# Display the last 5 rows of the Tesla Revenue DataFrame
print("\nLast 5 rows of Tesla Revenue DataFrame:")
print(tesla_revenue.tail())


Last 5 rows of Tesla Revenue DataFrame:
         Date Revenue
48 2010-09-30      31
49 2010-06-30      28
50 2010-03-31      21
52 2009-09-30      46
53 2009-06-30      27


## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [15]:
gme = yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


In [16]:
gme_data = gme.history(period="max")
gme_data.head()

,Open,High,Low,Close,Volume,Dividends,Stock Splits
Date,,,,,,,
2002-02-13 00:00:00-05:00,1.620129,1.693350,1.603296,1.691667,76216000,0.0,0.0
2002-02-14 00:00:00-05:00,1.712707,1.716074,1.670626,1.683250,11021600,0.0,0.0
2002-02-15 00:00:00-05:00,1.683251,1.687459,1.658002,1.674835,8389600,0.0,0.0
2002-02-19 00:00:00-05:00,1.666418,1.666418,1.578047,1.607504,7410400,0.0,0.0
2002-02-20 00:00:00-05:00,1.615921,1.662210,1.603296,1.662210,6892800,0.0,0.0


**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [17]:
gme_data.reset_index(inplace=True) #reset the index
gme_data.head()

,Date,Open,High,Low,Close,Volume,Dividends,Stock Splits
0,2002-02-13 00:00:00-05:00,1.620129,1.693350,1.603296,1.691667,76216000,0.0,0.0
1,2002-02-14 00:00:00-05:00,1.712707,1.716074,1.670626,1.683250,11021600,0.0,0.0
2,2002-02-15 00:00:00-05:00,1.683251,1.687459,1.658002,1.674835,8389600,0.0,0.0
3,2002-02-19 00:00:00-05:00,1.666418,1.666418,1.578047,1.607504,7410400,0.0,0.0
4,2002-02-20 00:00:00-05:00,1.615921,1.662210,1.603296,1.662210,6892800,0.0,0.0


## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data_2`.


In [18]:
url_gme = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html'
html_data_2 = requests.get(url_gme).text # Fetch the data from the URL


Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


In [19]:
soup_2 = BeautifulSoup(html_data_2, 'html5lib') # Parse the HTML data

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column.


> **Note: Use the method similar to what you did in question 2.**  


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [38]:
# Initialize an empty DataFrame to store the Tesla Revenue table
gme_revenue = pd.DataFrame(columns=['Date', 'Revenue'])

tables = soup_2.find_all('table') #finds all the tables in the HTML

for table in tables: #finds all the tables in the HTML
    if 'GameStop Quarterly Revenue' in str(table): #checks if the table contains the text 'GameStop Quarterly Revenue'
        print("Found the GameStop Quarterly Revenue table.")

        # Extract rows from the table
        for row in table.find("tbody").find_all("tr"):
            col = row.find_all("td")
            date = col[0].text.strip() # gets the date from the first column
            revenue = col[1].text.strip() # gets the revenue from the second column
    
            # Append the data to the DataFrame
            gme_revenue = pd.concat([gme_revenue, pd.DataFrame({"Date":[date],"Revenue":[revenue]})], ignore_index=True)
    
        # Clean the revenue data
        gme_revenue["Revenue"] = gme_revenue['Revenue'].str.replace(',|\$',"", regex=True)
        break # exit the loop after finding and processing the correct table

# Convert the 'Date' column to datetime format
gme_revenue['Date'] = pd.to_datetime(gme_revenue['Date'])

# Print the cleaned DataFrame
print(gme_revenue)

Found the GameStop Quarterly Revenue table.
         Date Revenue
0  2020-04-30    1021
1  2020-01-31    2194
2  2019-10-31    1439
3  2019-07-31    1286
4  2019-04-30    1548
..        ...     ...
57 2006-01-31    1667
58 2005-10-31     534
59 2005-07-31     416
60 2005-04-30     475
61 2005-01-31     709

[62 rows x 2 columns]


Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [39]:
# Display the last 5 rows of the GameStop Revenue DataFrame
print("\nLast 5 rows of GameStop Revenue DataFrame:")
print(gme_revenue.tail())


Last 5 rows of GameStop Revenue DataFrame:
         Date Revenue
57 2006-01-31    1667
58 2005-10-31     534
59 2005-07-31     416
60 2005-04-30     475
61 2005-01-31     709


## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`.

```
    
</details>


In [50]:
# Call the make_graph function
make_graph(tesla_data, tesla_revenue, 'Tesla')


C:\Users\jschi\AppData\Local\Temp\ipykernel_18808\2068661864.py:9: UserWarning:

The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.

C:\Users\jschi\AppData\Local\Temp\ipykernel_18808\2068661864.py:10: UserWarning:

The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.



## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`

```
    
</details>


In [40]:
make_graph(gme_data, gme_revenue, 'GameStop')

C:\Users\jschi\AppData\Local\Temp\ipykernel_18808\2068661864.py:9: UserWarning:

The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.

C:\Users\jschi\AppData\Local\Temp\ipykernel_18808\2068661864.py:10: UserWarning:

The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.



<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
